### ISSUE TITLE
visibleRows callback?

### ISSUE THREAD

**Issue by matrinox** [facebook/fixed-data-table#66]
We've got a table view that loads data as it is fetched. Using `rowGetter`, we check if there is data. If there is none that index is queued for fetching. Once it reaches a certain size or after a certain debounce time, the missing indexes are fetched.

That solution is serviceable at best but not at all efficient. It's not efficient when scrolling up a row (since it doesn't know direction) and when re-fetching the table (such as sorting or changing query) there isn't an easy way to find the starting offset to fetch from. Unless...

We have a `visibleRows` callback. That way we know exactly what is seen as each row is fetched. `10...20` then `11...21`, etc. It shows direction and lets us know the last visible rows to fetch when re-fetching.

tl;dr need way to know what rows are visible for smarter fetching of server data.

**Comment by pieterv** [comment]
Hey @matrinox, i have been thinking about this `visibleRows` style API for a while but i'm not totally convinced it would be needed. What would be the advantage to using this `visibleRows` callback over just batching up all data request within a single event loop and then breaking that into request batches? (other than the frame work providing the logic for you :P)

Here is an example of how i would build the batching logic: http://jsbin.com/temufa/9/edit?js,output
You will notice it will also request data outside of the view based on where it thinks you will scroll to, because we are using the rowGetter we can take advantage of the row prerendering triggering requesting data.

**Comment by hedgerwang** [comment]
@matrinox : this sounds more like a issue with data fetching and processing that the data loader (store) should optimize for.

One way to address this issue is to break the rows into chunks, such as [0-50], [51-100], ...etc. The store will does all the data caching and invoke the data fetching action if the chunk referred hasn't been fetched or loading yet.

the table is just a view, after all.

**Comment by matrinox** [comment]
@hedgerwang Good idea, separating the concerns

But what if you click on the scroll bar to, I don't know, line 49-59? Do you fetch [0-49] and [50-99]? And then what happens when you refresh the table view, do you do another 2 fetches?

I suppose it could work in the meantime but long term I think that is the best solution

**Comment by hedgerwang** [comment]
@matrinox: It's basically like how the normal pagination would work.

Depends on the network and server resources, you breakdown the data (rows) into chunks then the scrollbar is just a tool for you to move between "pages".

If user happen to step into the position that require fetching more than one page, so be it or you can solve that with batched requested.

The proposed API `visibleRows` does not help to ease the problem because the visible rows are constantly changing while scrolling and you'd still need to implement batching and pagination logics for your data-fetching library eventually for perf sake.

**Comment by pieterv** [comment]
Agree with @hedgerwang  but this does seem initially non obvious. I think this would be a good example we could add to the site. Im going to close this out and create a task for it.

**Comment by matrinox** [comment]
:+1:
